kyuubi-common/src/main/scala/org/apache/kyuubi/operation/log/OperationLog.scala

```diff
@@ -144,10 +156,29 @@ class OperationLog(path: Path) {
       logs.addAll(extraLogs)
     }
 
-    val tColumn = TColumn.stringVal(new TStringColumn(logs, ByteBuffer.allocate(0)))
-    val tRow = new TRowSet(0, new JArrayList[TRow](logs.size()))
-    tRow.addToColumns(tColumn)
-    tRow
+    toRowSet(logs)
+  }
+
+  def read(from: Int, size: Int): TRowSet = synchronized {
+    var pos = from
+    if (pos < 0) {
+      // just fetch forward
+      pos = lastReadPos
+    }
+    if (seekableReader == null) {
+      seekableReader = new SeekableBufferedReader(Seq(path) ++ extraPaths)
+    } else {
+      // if from < last pos, we should reload the reader
+      // otherwise, we can reuse the existed reader for better performance
+      if (pos < lastReadPos) {
+        seekableReader.close()
+        seekableReader = new SeekableBufferedReader(Seq(path) ++ extraPaths)
+      }
+    }
+
+    val it = seekableReader.readLine(pos, size)
+    lastReadPos = pos + size
```

Comment:
the it.size may be small than size.

```
    lastReadPos = pos + it.size
```

Reply:
good point, use materialized list size instead of the required size